File: clients/dotnet-NewComponent-BrokerTCP/Tests/SimpleTestsFramework/Step.cs
using System;
using System.Threading;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Tests.SimpleTestsFramework
{
    /// <summary>
    /// Represents a step in a test.
    /// </summary>
    public abstract class Step
    {

        readonly private string name;
        volatile private string reasonForFailure;
        readonly private ManualResetEvent manualResetEvent;
        volatile private System.Action run;

        volatile private bool done = false;
        volatile private bool sucess = false;

        public ManualResetEvent ManualResetEvent
        {
            get { return manualResetEvent; }
        }

        public Step(String name)
        {
            this.name = name;
            manualResetEvent = new ManualResetEvent(false);
        }

        public String Name
        {
            get { return name; }
        }

        public bool Done
        {
            get { return done; }
            set { 
                this.done = value;
                manualResetEvent.Set();
            }
        }

        public bool Sucess
        {
            get { return sucess; }
            set { this.sucess = value; }
        }

        public bool IsFailure()
        {
            return (Done) && (!Sucess);
        }

        public string ReasonForFailure
        {
            set
            {
                Sucess = false;
                this.reasonForFailure = value;
                Done = true;
            }
            get
            {
                return this.reasonForFailure;
            }
        }

        public Step Call()
        {
            if (IsFailure())
            {
                return this;
            }
            try
            {
                if (run != null)
                    run();
            }
            catch (Exception e)
            {
                this.ReasonForFailure = "Exception - " + e.Message;
            }

            return this;
        }

        public System.Action Runnable
        {
            get { return run; }
            set { run = value; }
        }
    }
}

File: test/broker/src/main/java/pt/com/broker/functests/Step.java
package pt.com.broker.functests;

import java.util.concurrent.Callable;

public abstract class Step implements Callable<Step>
{
	private String name;
	private String reasonForFailure;

	private boolean done = false;
	private boolean sucess = false;

	public Step(String name)
	{
		this.name = name;
	}

	public String getName()
	{
		return name;
	}

	protected void setDone(boolean done)
	{
		this.done = done;
	}

	public boolean isDone()
	{
		return done;
	}

	protected void setSucess(boolean sucess)
	{
		this.sucess = sucess;
	}

	public boolean isSucess()
	{
		return sucess;
	}

	public boolean isFailure()
	{
		return (done) && (!sucess);
	}

	public void setReasonForFailure(String reasonForFailure)
	{
		setDone(true);
		setSucess(false);
		this.reasonForFailure = reasonForFailure;
	}

	public String getReaseonForFailure()
	{
		return reasonForFailure;
	}

	@Override
	public final Step call() throws Exception
	{
		if (isFailure())
			return this;

		return run();
	}

	public abstract Step run() throws Exception;
}
